platform-sdk/swirlds-common/src/main/java/com/swirlds/common/wiring/counters/BackpressureBlocker.java

```diff
@@ -0,0 +1,90 @@
+/*
+ * Copyright (C) 2023 Hedera Hashgraph, LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.swirlds.common.wiring.counters;
+
+import static java.util.concurrent.TimeUnit.NANOSECONDS;
+
+import edu.umd.cs.findbugs.annotations.NonNull;
+import java.util.Objects;
+import java.util.concurrent.ForkJoinPool.ManagedBlocker;
+import java.util.concurrent.atomic.AtomicLong;
+
+/**
+ * This class is used to implement backpressure in a {@link java.util.concurrent.ForkJoinPool} friendly way.
+ */
+class BackpressureBlocker implements ManagedBlocker {
+
+    /**
+     * The current count. This object will attempt to block until this count can be increased by one without exceeding
+     * the maximum capacity.
+     */
+    private final AtomicLong count;
+
+    /**
+     * The maximum desired capacity. It is possible that the current count may exceed this capacity (i.e. if
+     * {@link ObjectCounter#forceOnRamp()} is used to bypass the capacity).
+     */
+    private final long capacity;
+
+    /**
+     * The amount of time to sleep while waiting for capacity to become available, or 0 to not sleep.
+     */
+    private final long sleepNanos;
```

Comment:
which thread does the sleeping?

Reply:
If by thread you mean the OS level concept, which thread is not really defined. The fork join pool may do this on any thread it chooses, and may even change threads in the middle of the block.

From the logical perspective, the blocking behavior IS experienced by the caller/inserter. If scheduler `A` wants to push a task into the logical queue for scheduler `B` but `B` is at maximum capacity, `A` will be forced to wait, and this mechanism will enforce that wait.